crates/dids/src/methods/dht/document_packet/service.rs

```diff
@@ -0,0 +1,237 @@
+            .expect("Failed to convert ResourceRecord to Service");
+
+        assert_eq!(service, service2);
+    }
+
+    #[test]
+    fn test_to_record_resource_service_id_multiple_sharps() {
+        // The URL spec seems to say that multiple "#" is invalid,
+        // but after trying some actual URL implementations,
+        // this behavior is more common.
+        let did_uri = "did:dht:123";
+        let id = "did:dht:123#hey#ya"; // multiple "#"
+        let r#type = "some_type";
+        let service_endpoint = "foo.tbd.website";
+        let service = Service {
+            id: id.to_string(),
+            r#type: OneOrMany::One(r#type.to_string()),
+            service_endpoint: OneOrMany::One(service_endpoint.to_string()),
+        };
+
+        let resource_record = service
+            .to_resource_record(0)
+            .expect("Expected to create resource record from service");
+        let service2 = Service::from_resource_record(did_uri, resource_record).expect("msg");
+        assert_eq!(service, service2);
+    }
+
+    #[test]
+    fn test_to_record_resource_missing_fragment() {
+        let did_uri = "did:dht:123"; // missing "#0"
+        let r#type = "some_type";
+        let service_endpoint = "foo.tbd.website";
+        let service = Service {
+            id: did_uri.to_string(),
+            r#type: OneOrMany::One(r#type.to_string()),
+            service_endpoint: OneOrMany::One(service_endpoint.to_string()),
+        };
+
+        let resource_record = service
+            .to_resource_record(0)
+            .expect_err("Expected error due to service.id missing fragment");
+        match resource_record {
+            DocumentPacketError::MissingFragment(id) => {
+                assert_eq!(id, did_uri)
+            }
+            _ => panic!(),
+        }
+    }
+
+    #[test]
+    fn test_from_record_resource_rdata_not_txt() {
+        let name = Name::new_unchecked(&format!("_s{}._did", 0)).into_owned();
+
+        let resource_record = ResourceRecord::new(
+            name,
+            simple_dns::CLASS::IN,
+            DEFAULT_TTL,
+            RData::A(A { address: 0 }), // not RData::TXT
+        );
+
+        let error = Service::from_resource_record("did:ex:abc", resource_record).expect_err("");
+        match error {
+            DocumentPacketError::RDataError(_) => {}
+            _ => panic!(),
+        };
+    }
+
+    #[test]
+    fn test_from_record_resource_rdata_txt_malformed() {
+        let name = Name::new_unchecked(&format!("_s{}._did", 0)).into_owned();
+
+        let txt = TXT::new().with_string("a=b=c;;;").unwrap();
+        let resource_record = ResourceRecord::new(
+            name,
+            simple_dns::CLASS::IN,
+            DEFAULT_TTL,
+            RData::TXT(txt), // Not ';' separated entries
+        );
+
+        let error = Service::from_resource_record("did:ex:abc", resource_record).expect_err("");
```

Comment:
```suggestion
        let error = Service::from_resource_record("did:ex:abc", resource_record).expect_err("Expected error because malformed txt");
```